lib/faraday/error.rb

```diff
@@ -79,25 +79,62 @@ def exc_msg_and_response!(exc, response = nil)
 
     # Pulls out potential parent exception and response hash.
     def exc_msg_and_response(exc, response = nil)
-      if exc.is_a?(Exception)
+      case exc
+      when Exception
         [exc, exc.message, response]
-      elsif exc.is_a?(Hash) || exc.is_a?(Faraday::Env)
-        http_status = exc.fetch(:status)
+      when Hash
+        [nil, build_error_message_from_hash(exc), exc]
+      when Faraday::Env
+        [nil, build_error_message_from_env(exc), exc]
+      else
+        [nil, exc.to_s, response]
+      end
+    end
 
-        request = exc.fetch(:request, nil)
+    private
 
-        if request.nil?
-          exception_message = "the server responded with status #{http_status} - method and url are not available " \
-                              'due to include_request: false on Faraday::Response::RaiseError middleware'
-        else
-          exception_message = "the server responded with status #{http_status} for #{request.fetch(:method).upcase} " \
-                              "#{request.fetch(:url)}"
-        end
+    def build_error_message_from_hash(hash)
+      # Be defensive with external Hash objects - they might be missing keys
+      status = hash.fetch(:status, nil)
+      request = hash.fetch(:request, nil)
 
-        [nil, exception_message, exc]
-      else
-        [nil, exc.to_s, response]
+      return fallback_error_message(status) if request.nil?
+
+      method = request.fetch(:method, nil)
+      url = request.fetch(:url, nil)
+      build_status_error_message(status, method, url)
+    end
+
+    def build_error_message_from_env(env)
+      # Faraday::Env is internal - we can make reasonable assumptions about its structure
+      status = env.status
+
+      # Check if we have request data in nested :request key (RaiseError middleware format)
+      request = env.fetch(:request, nil)
+      if request
```

Comment:
```suggestion
      if (request = env.fetch(:request, nil))
```

Do you like/don't like that format?

Reply:
I do! Thanks for the suggestion